### PR TITLE
Improved Linux platform support & some tweaks/fixes

### PR DESCRIPTION
**Pull request by ximion**
Hi!
Thank you for writing this software! I originally just wanted to include improved Linux support in this PR, but picked up a few other changes along the way. This PR changes the following:
 * QMake now uses [pkg-config](https://en.wikipedia.org/wiki/Pkg-config) on Linux to find OpenCV which pretty much means you'll be able to instantly build the software on Linux without any include/link path tweaking as long as OpenCV is installed
 * The main window will now be centered on screen (it bothered me to always have all windows right at the top left screen edge)
 * The V4L backend is preferred on Linux, DirectShow is preferred on Windows (based on my own experience with GStreamer on Linux not liking the older DAQ board firmware - the new one doesn't appear to have the same issue, but better be safe. Usually V4L will be picked up automatically anyway) The app also won't even try DirectShow on Linux now, as that will never be found anyway.
 * Don't try to record if the user has forgotten to set a data base directory, otherwise the software will crash or write to arbitrary places (this may have happened to me after mistyping the JSON key... ^^)
 * Various crash fixes if the user managed to request nonsensical actions via the UI

I hope you'll find this useful! Please check if the Windows build still works - as far as I know, the WIN32 condition also covers 64-bit Windows, so Windows builds should be unaffected by this change, but better be sure.
Cheers,
    Matthias

Edit: Since this application is using QML, and QML is known to be a bit cumbersome on desktops and bigger screens, it may be interesting for you to look into [Kirigami](https://kde.org/products/kirigami/) if you don't know that already and have a new project that uses QML. Kirigami is a responsive QML component framework providing basic types and shapes that integrate well with the desktop environment or mobile UI and is available on all major platforms.

### CODE REVIEW

**Comment by daharoni** [comment]
Thanks so much for the pull request @ximion! I'll look through it tonight and also validate that it runs on Windows. 

Quick question: @phildong had previously done a quick port of the software to Linux and noticed that the software was not able to control the parameters of the Miniscope (adjust LED brightness, set gain, ...). Were you able to test this and was everything working? This software, and compatible Miniscope DAQ firmware, handles control of the Miniscope very differently than the previous version of software/firmware. Wondering if you found any issues with this?

**Comment by ximion** [comment]
Hi @daharoni :-) I can reproduce @phildong 's issues here as well, and poked the system a bit to figure out why this wasn't working.
I can's say with 100% confidence yet that I found the culprit, but I have a theory... I was using Wireshark to intercept the USB data blocks sent by the computer, and when attempting to send the GAMMA and HUE properties it would always send the same value. Sending CONTRAST however would send the correct value. A different webcam works fine though and HUE values are submitted properly to the device.

Looking into what the device sends via the V4L Test Bench it appears that controls for these values are indeed messed up there as well:
grafik
(there should be sliders available, and the values do not go past 9 (decimal) which isn't right)

Running the V4L compliance test on the MINISCOPE DAQ device with the new firmware reveals the following:
```
v4l2-compliance SHA: not available, 64 bits

Compliance test for uvcvideo device /dev/video0:

Driver Info:
        Driver name      : uvcvideo
        Card type        : MINISCOPE: MINISCOPE
        Bus info         : usb-0000:00:14.0-2
        Driver version   : 5.4.19
        Capabilities     : 0x84a00001
                Video Capture
                Metadata Capture
                Streaming
                Extended Pix Format
                Device Capabilities
        Device Caps      : 0x04200001
                Video Capture
                Streaming
                Extended Pix Format
Media Driver Info:
        Driver name      : uvcvideo
        Model            : MINISCOPE: MINISCOPE
        Serial           : 
        Bus info         : usb-0000:00:14.0-2
        Media version    : 5.4.19
        Hardware revision: 0x00000000 (0)
        Driver version   : 5.4.19
Interface Info:
        ID               : 0x03000002
        Type             : V4L Video
Entity Info:
        ID               : 0x00000001 (1)
        Name             : MINISCOPE: MINISCOPE
        Function         : V4L2 I/O
        Flags         : default
        Pad 0x01000007   : 0: Sink
          Link 0x02000010: from remote pad 0x100000a of entity 'Extension 3': Data, Enabled, Immutable

Required ioctls:
        test MC information (see 'Media Driver Info' above): OK
        test VIDIOC_QUERYCAP: OK

Allow for multiple opens:
        test second /dev/video0 open: OK
        test VIDIOC_QUERYCAP: OK
        test VIDIOC_G/S_PRIORITY: OK
        test for unlimited opens: OK

Debug ioctls:
        test VIDIOC_DBG_G/S_REGISTER: OK (Not Supported)
        test VIDIOC_LOG_STATUS: OK (Not Supported)

Input ioctls:
        test VIDIOC_G/S_TUNER/ENUM_FREQ_BANDS: OK (Not Supported)
        test VIDIOC_G/S_FREQUENCY: OK (Not Supported)
        test VIDIOC_S_HW_FREQ_SEEK: OK (Not Supported)
        test VIDIOC_ENUMAUDIO: OK (Not Supported)
        test VIDIOC_G/S/ENUMINPUT: OK
        test VIDIOC_G/S_AUDIO: OK (Not Supported)
        Inputs: 1 Audio Inputs: 0 Tuners: 0

Output ioctls:
        test VIDIOC_G/S_MODULATOR: OK (Not Supported)
        test VIDIOC_G/S_FREQUENCY: OK (Not Supported)
        test VIDIOC_ENUMAUDOUT: OK (Not Supported)
        test VIDIOC_G/S/ENUMOUTPUT: OK (Not Supported)
        test VIDIOC_G/S_AUDOUT: OK (Not Supported)
        Outputs: 0 Audio Outputs: 0 Modulators: 0

Input/Output configuration ioctls:
        test VIDIOC_ENUM/G/S/QUERY_STD: OK (Not Supported)
        test VIDIOC_ENUM/G/S/QUERY_DV_TIMINGS: OK (Not Supported)
        test VIDIOC_DV_TIMINGS_CAP: OK (Not Supported)
        test VIDIOC_G/S_EDID: OK (Not Supported)

Control ioctls (Input 0):
                fail: v4l2-test-controls.cpp(76): def < min || def > max
                fail: v4l2-test-controls.cpp(208): invalid control 00980900
        test VIDIOC_QUERY_EXT_CTRL/QUERYMENU: FAIL
                fail: v4l2-test-controls.cpp(318): node->controls.find(qctrl.id) == node->controls.end()
        test VIDIOC_QUERYCTRL: FAIL
        test VIDIOC_G/S_CTRL: OK
                warn: v4l2-test-controls.cpp(715): g_ext_ctrls(0) invalid error_idx 0
        test VIDIOC_G/S/TRY_EXT_CTRLS: OK
        test VIDIOC_(UN)SUBSCRIBE_EVENT/DQEVENT: OK (Not Supported)
        test VIDIOC_G/S_JPEGCOMP: OK (Not Supported)
        Standard Controls: 0 Private Controls: 0

Format ioctls (Input 0):
        test VIDIOC_ENUM_FMT/FRAMESIZES/FRAMEINTERVALS: OK
                warn: v4l2-test-formats.cpp(1327): S_PARM is supported but doesn't report V4L2_CAP_TIMEPERFRAME
                fail: v4l2-test-formats.cpp(1229): node->has_frmintervals && !cap->capability
        test VIDIOC_G/S_PARM: FAIL
        test VIDIOC_G_FBUF: OK (Not Supported)
                fail: v4l2-test-formats.cpp(341): !colorspace
                fail: v4l2-test-formats.cpp(448): testColorspace(node->has_inputs || node->has_outputs, pix.pixelformat, pix.colorspace, pix.ycbcr_enc, pix.quantization)
        test VIDIOC_G_FMT: FAIL
                fail: v4l2-test-formats.cpp(341): !colorspace
                fail: v4l2-test-formats.cpp(448): testColorspace(node->has_inputs || node->has_outputs, pix.pixelformat, pix.colorspace, pix.ycbcr_enc, pix.quantization)
        test VIDIOC_TRY_FMT: FAIL
                fail: v4l2-test-formats.cpp(341): !colorspace
                fail: v4l2-test-formats.cpp(448): testColorspace(node->has_inputs || node->has_outputs, pix.pixelformat, pix.colorspace, pix.ycbcr_enc, pix.quantization)
        test VIDIOC_S_FMT: FAIL
        test VIDIOC_G_SLICED_VBI_CAP: OK (Not Supported)
        test Cropping: OK (Not Supported)
        test Composing: OK (Not Supported)
        test Scaling: OK (Not Supported)

Codec ioctls (Input 0):
        test VIDIOC_(TRY_)ENCODER_CMD: OK (Not Supported)
        test VIDIOC_G_ENC_INDEX: OK (Not Supported)
        test VIDIOC_(TRY_)DECODER_CMD: OK (Not Supported)

Buffer ioctls (Input 0):
        test VIDIOC_REQBUFS/CREATE_BUFS/QUERYBUF: OK
        test VIDIOC_EXPBUF: OK
        test Requests: OK (Not Supported)

Total for uvcvideo device /dev/video0: 45, Succeeded: 39, Failed: 6, Warnings: 2
```
The failures in the `Control ioctls` section are what's relevant, they tell that the device apparently sets a default value not in the min/max value range it has for integers.

So, at the moment I think what happens is that the userspace tool tries to set a higher value for the particular properties, but the driver and/or interface libraries clamp it to the minimum/maximum values specified, so the intended value never reaches the device and no change of the property is possible.
In other words, I think this is a firmware bug (likely not a driver bug because the same driver works for other devices and the old Miniscope firmware). The think that makes me doubt in the theory is that this stuff apparently works on Windows, which is really odd unless the Windows driver is ignoring device limits.
But the firmware is where I would need to look next.

FWIW, the new device firmware is pretty stable - usually my DAQ board has to be reconnected at least every 20min because something goes wrong and it can't deliver frames anymore. Now, with the new firmware, I had a really long debugging session with the device connected for hours (I also left for food ^^) and there wasn't even a single connection failure :-)
This is really encouraging!

**Comment by ximion** [comment]
Some more evidence for the broken firmware theory:
```
Streaming Parameters Video Capture:
        Capabilities     : timeperframe
        Frames per second: 30.000 (30/1)
        Read buffers     : 0
                     brightness 0x00980900 (int)    : min=0 max=-1 step=1 default=0 value=0
                       contrast 0x00980901 (int)    : min=0 max=65535 step=1 default=0 value=1208
                     saturation 0x00980902 (int)    : min=0 max=65535 step=1 default=0 value=0
                            hue 0x00980903 (int)    : min=0 max=-1 step=1 default=0 value=0
```
This was querying the device controls directly, a max value of -1 is definitely wrong and probably an indicator for an error.
And indeed, when tracing the uvcvideo driver:
```
kernel: uvcvideo: Suspending interface 1
kernel: uvcvideo: Suspending interface 0
kernel: uvcvideo: uvc_v4l2_open
kernel: uvcvideo: Resuming interface 0
kernel: uvcvideo: Resuming interface 1
kernel: uvcvideo: Failed to query (GET_DEF) UVC control 12 on unit 2: -32 (exp. 4).
kernel: uvcvideo: Control error 6
kernel: uvcvideo: Failed to query (GET_DEF) UVC control 12 on unit 2: -32 (exp. 4).
kernel: uvcvideo: Control error 6
kernel: uvcvideo: uvc_v4l2_release
kernel: uvcvideo: Suspending interface 1
kernel: uvcvideo: Suspending interface 0
```
Presumably those query errors, which I initially though were not relevant, actually are relevant. I couldn't find anything obviously wrong when quickly looking over the firmware though (maybe someone more knowledgeable in UVC can look into that as well...) It's weird that Windows apparently doesn't care, still.

**Comment by ximion** [comment]
I've found the problem... Or rather: A problem. I will follow up in a bug report against the firmware, since this is a bit tricky to fix.
The main issue is that the firmware is abusing values designated to be signed integers as unsigned integers, which leads to chaos. I still have no idea why this works on Windows, but presumable the Windows driver does way less error checking and value sanitization.

**Comment by daharoni** [comment]
Sorry it took me this long to get back to this thread @ximion. 

I think part of the issue might be that different OS's camera streaming backends have different constraints on the min/max of certain camera parameters (brightness, gain, ...). When I was first working on this I found a direct show webpage (which for the life of me I cannot find right now) which gave the hardset range of allowable values for different camera properties. 

From what I worked out, OpenCV's VidoeCapture::set function takes in a double as a value and sends it to backend functions. For the Direct Show backend, there are only certain camera parameters which support values of 0 to 2^16-1. Some other parameters only support values between -1 and 1. Since the new DAQ software/firmware is using these parameters as a way to just send bytes of data to the DAQ, if the Linux backend doesn't support the same range of values this could cause an issue.

**Comment by daharoni** [comment]
Oh, I just found the DirectShow webpage I mentioned in my previous post. https://docs.microsoft.com/en-us/previous-versions/ms787924(v%3Dvs.85)

I also just read your comments and pull request on the DAQ firmware repo and I think we are discussing basically the same thing. It doesn't seem like directshow has much of an issue with the max value of a camera properties being set (-1) to below the min value. When I was first working on this I basically just went through all possible properties and tried to see the range of values they allowed. Since I already had to do some, not so kosher, type casting between doubles and ints I didn't pay much attention to other details that weren't causing issues. In the end, all the properties used in the firmware (saturation, hue, gain, brightness, contrast, gamma, and sharpness) were able to handle 16bit values in and out without error on Windows.

**Comment by ximion** [comment]
> Sorry it took me this long to get back to this thread

Don't worry, usually I am the one who is slow to reply ^^ And I am not in a rush with this effort at all (it will become a high priority though once we have a v4 Miniscope, as the Miniscope is integrated in a DAQ pipeline that fully runs on Linux and the v4 scope needs the new firmware...)

I actually find this issue quite fascinating, because I can inspect every piece of code run in Linux, including the driver part, but I have no idea what's going on inside Windows - and still by inference, you actually learn a lot about the other OS and how hardware actually interacts with the computer. It's a bit like solving a (so far, fun) puzzlebox.

This probably should go into https://github.com/Aharoni-Lab/Miniscope-DAQ-Cypress-firmware/pull/1, but the "issue" here is that the Miniscope hardware identifies itself as UVC device, which is part of the USB specification. 

You can read the spec here: https://www.usb.org/document-library/video-class-v15-document-set (the "UVC 1.5 Class specification" document is relevant here, it defines much of the communication protocol for control values - also, this is not at all news for the guy who created the hardware, but maybe that's interesting for others finding this issue report in the future).

The Windows driver apparently just passes bytes through for control values and doesn't care at all about what they are, while the Linux driver is actually sanitizing values to match the USB UVC spec exactly. I adjusted the firmware to follow the spec, but it still doesn't work correctly...

> For the Direct Show backend, there are only certain camera parameters which support values of 0 to 2^16-1. Some other parameters only support values between -1 and 1. Since the new DAQ software/firmware is using these parameters as a way to just send bytes of data to the DAQ, if the Linux backend doesn't support the same range of values this could cause an issue.

> Oh, I just found the DirectShow webpage I mentioned in my previous post. https://docs.microsoft.com/en-us/previous-versions/ms787924(v%3Dvs.85)

Oooooh yes indeed! Thanks a lot for that, and especially for the DirectShow doc page! That may help immensely. Do you have any documentation on which values the hardware actually expects to arrive to trigger actions? (it's in the JSON files and the code, but that's harder to follow) Is the information in the JSON files exactly what the hardware expects to see, or is any of that adjusted for some kind of value-mangling that DirectShow may perform?

Even though the UVC spec mandates some values to be signed integers (like hue) they are still 16bit values, so your byte-transmission should in theory work just as well on Linux as it does on Windows.

As for casting from/to double: Yeah, that's not super clean, but also inevitable when working with OpenCV. I wonder why they picked doubles in the first place - I worked with a bunch of video capture APIs the past week, and not a single one actually wanted floats for any control value, which makes sense since those suck to deal with in most hardware. Presumably there is one backend OpenCV supports that does need floats, or someone just wanted to futureproof the API by going with a very generic datatype. But in that case they should either have sanitized the input values to a common range, or have added API to query min/max values of a camera property.

**Comment by daharoni** [comment]
Hey @ximion,

> Do you have any documentation on which values the hardware actually expects to arrive to trigger actions? (it's in the JSON files and the code, but that's harder to follow) Is the information in the JSON files exactly what the hardware expects to see, or is any of that adjusted for some kind of value-mangling that DirectShow may perform?
I haven't written up documentatino on what the hardware actually expects but here is the general workflow:

1. The JSON file holds information for each camera property including how the control is displayed and the relevant data that will be sent to the DAQ when the control is interacted with.
2. The QML file for each control type is able to modify the information contained in the JSON file a bit before it get sent over to the C++ end to be sent to the DAQ.
3. The backend then sends a packet of data over UVC camera property values to the DAQ. The structure of this packet is the same for all controls and camera properties and can be up to 6 bytes long. Once the DAQ receives a data packet, it adds it to a queue to be sent over I2C at the end of the next frame acquisition.

The goal of all of this is to remove the need to do any hardcoding of I2C commands in the DAQ and allow everything to be configured in the DAQ software (and ideally just in the JSON file without the need to ever jump into the QML or C++ code).

It has turned out to work pretty well and has allowed us to quickly add new components and imaging sensors without all the added work of modifying firmware and software that we had to do for the previous version. I would say the biggest downside currently is every once in a while we add a new component to the Miniscope that needs to talk over I2C and the interaction between the GUI control for this component and the data that needs to be sent to it is a bit different than previous devices. I then have to get a little creative in the JSON and QML to make this all work.
